### ISSUE TITLE
Wrong inference for pattern matching with custom tuple subclass with literal pattern

### ISSUE THREAD

**Issue by lkct** [microsoft/pyright#8967]
**Describe the bug**
Better illustrated with the code below.

My understanding of [8.6.4.8. Sequence Patterns](https://docs.python.org/3/reference/compound_stmts.html#sequence-patterns) is that:
- `mt` is a sequence (subclass of tuple), so it can be matched to a sequence pattern;
- It has an unknown length, so it may match patterns of any length from the view of a static analyser;
- Also unknown length means it may miss any fixed-length pattern.

Therefore in the following example all cases can be matched.

---

However, the actual behaviour is that: the second `reveal_type(b)` gives no output; mouse hovering in vscode suggests `b: Never`. It appears that case 1 shadows case 2 of a different length, while case 3 with the same length still works.

Furthermore, the return type of foo2 is inferred to be `Literal['1', '3']` but not `None` (no case matched, no explicit return), meaning the length of mt is inferred to be 4, so that case 3 captures everything (effectively irrefutable).

Ablation study: builtin tuple works fine; `MyTuple` without a literal `'a'` in the patterns is also fine. Please refer to the playground with longer code linked below.

**Code or Screenshots**
```python
class MyTuple(tuple[str, ...]):
    def __new__(cls, iterable: Iterable[Any]):
        return super().__new__(cls, (str(item) for item in iterable))

def foo2():
    mt = MyTuple("abcde")
    match mt:
        case ("a", b, _, _):
            reveal_type(b)  # str
            return "1"

        case ("a", b, _, _, _):
            reveal_type(b)  # no output
            return "2"

        case (_, b, _, _):
            reveal_type(b)  # str
            return "3"

        case (_, b, _, _, _):
            reveal_type(b)  # no output
            return "4"


reveal_type(foo2())  # Literal['1', '3']
```

For more cases, see code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgElUkRjkAoUSKVATwvSwDmZStVoBBLNwA0UEEgBuSAvAD6PCklZaY8AjhxQAstwAqAVwrIAFKgvIA2jlQgZeNwF0AlAC5WUf1AAJkjAUCoqWEgA7uFWOjgy6IzMRMje9MksSPYS3F6%2BAYWySLYgWFA4FkxWnnjhkTEqcfAJUFZOIFZJSBCeUMDUZIzQ-ENMWZ6eWqzBoQNgAIw1BQG0ALxcdkhWAETEMME7U4VyispqvNuoxwEQBKgwABZcK0WwekhtezsyRDIq-x8fjeRVOSlU6m2RBuIICclK5R2Cx2WlhMA%2BXwIPygfzC-0Br1hxTOEMuVmhwKJxQRUB2ACYUVowedIVZ5ktJtNZv0wGA6ctKVAIOsjKYtrt9ocYUL7k8hahCYV0ThPhLsbiAWEgVT-MzSRpydLYfCzGVacjUSDlarvr98XitYqQXqLgaKTrqabEQzLW9rW1NRqCYLjQpwa6oUbnSUvbSAMyMtEYqyB%2B2alTanUu1nurMxs07AAsjNY2bJ835nK03PmcYFhWFUA2xnMlm2exIUsFdwez2FTv8-pTdodGYHoLDLLJuapJoLFpDASHqdHwY9ZbdUbec%2B9JY321rNSm2l0%2BlFABl0E4rPAr6hHM58oLueF%2BElYir4MBEplUkh0gw4x-jkkhPrClQaJ0tSvlg75NB0XTDH0AwgGMIzlN0KTIMeMwhDiBAgByhLwE2UC3teHYHEgRyCvuN7Sj2crwOOQ62jiaaZrOk76pGi7bvmiILkmKqYuqaZrnmJIRoafETjS9J7tx0lEARHI4dyKkgJWhIQCRzbcJeFGStRDGyn2zGyaxWIjumnFEnRM72QJ5qJlayZsUGo52aGUk5lucmxgp0wOQRlZTEAA)

**VS Code extension or command-line**
Latest pyright (1.1.380) as provided by playground

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.381.
